Makefile

```diff
@@ -51,18 +51,26 @@ _output/share/lima/lima-guestagent.Linux-aarch64:
 	chmod 644 $@
 
 .PHONY: install
+install: | $(abspath $(DESTDIR)/$(PREFIX))
+install: DEST = $(realpath $(DESTDIR)/$(PREFIX))
```

Comment:
Since setting `DEST` is repeated on the `uninstall` target, I think it should instead be set globally at the top, just after the line providing the default for `PREFIX`.

Reply:
I can do it, and I agree that simple is better :)

But it would not work for a destination that does not already exist, because `realpath` returns an empty string in that case

Reply:
What we can do is: move DEST to the top scope but using `abspath` (which does not return an empty string for directories that do not already exist), similar to what I was doing below in the target prerequisite.

Reply:
> What we can do is: move DEST to the top scope but using `abspath`

I realize the original intent of this PR was to use `realpath` to check if the path exists, and fail if it doesn't. But now that the path is created when it doesn't exist, is there any reason to call `abspath`? Does anything user-observable work differently?

Sorry for getting pretty close to nit-picking territory now, but if the call doesn't serve any functional purpose, and doesn't cause any user-visible change in behaviour, then maybe it should be removed too? Because code that does not exist cannot have bugs. 😄 

And thinking about it some more, `abspath` may actually introduce bugs because it can generate the wrong path when symlinks and `../` path segments are involved:

```console
$ grep -A1 abspath: Makefile
abspath:
	@echo $(DEST)
$ ln -s /usr/local ~/foo
$ export DESTDIR=~/foo
$ export PREFIX=../lib
$ make abspath
/Users/jan/lib
$ (cd $DESTDIR/$PREFIX;pwd)
/usr/lib
```

So now I think the `abspath` call should be removed unless I've missed some reason it is useful/required.

Reply:
Hello Jan,

yes, the scope of the PR was that one.

The reason I used `abspath` is the same I used `realpath` in the first iteration of this PR.

Ie., to make the process of staged installs (via `DESTDIR`) more robust and clear (and actually work) w.r.t. to trailing slashes, etc.

This is what the user will see if we remove `abspath`:

```diff
diff --git a/Makefile b/Makefile
index 0b74f88..8e36c16 100644
--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 # Files are installed under $(DESTDIR)/$(PREFIX)
 PREFIX ?= /usr/local
-DEST := $(abspath $(DESTDIR)/$(PREFIX))
+DEST := $(DESTDIR)/$(PREFIX)

 GO ?= go
```

Outputs:

```console
DESTDIR=/tmp/ make install
mkdir -p "/tmp///usr/local"
cp -av _output/* "/tmp///usr/local"
_output/bin -> /tmp///usr/local/bin
...
```

If this is what you want, I'll do it. Even if do not totally like it atm :)

Reply:
Hi Leo,

> If this is what you want, I'll do it.

I'm only providing code review feedback. If you disagree, please push back! In the end it depends on what @AkihiroSuda wants.

As for the multiple slashes, isn't that a purely cosmetic issue? POSIX pathname definition says:

> Multiple successive &lt;slash> characters are considered to be the same as one &lt;slash>, except for the case of exactly two leading <slash> characters.

So paths starting with two leading slashes are potentially a concern, but I believe neither Linux nor macOS implement any special semantics in this case.

I do understand where you are coming from, as I'm always annoyed that e.g. `find /etc/ ...` shows names as `/etc//...` (and `find /etc ...` doesn't work because `/etc` is a symlink to `private/etc` on macOS).

If you are purely concerned about the successive slashes, maybe a strict textual substitution is preferable:

```Makefile
DEST := $(patsubst %/,%,$(DESTDIR))/$(patsubst /%,%,$(PREFIX))
```

But again this introduces subtle bugs, in this case if the paths contain consecutive whitespace:

```console
$ make abspath DESTDIR="/foo bar  baz/xyzzy/"
/foo bar baz/xyzzy/usr/local
```

So my personal preference would be to favour correctness over aesthetics and accept the untidy output, especially since this is about `make install` and not related to using `lima`.

Let's wait for @AkihiroSuda to make a decision.

Reply:
> this introduces subtle bugs, in this case if the paths contain consecutive whitespace

This is also true for `DEST := $(DESTDIR)/$(PREFIX)`; the right hand side must be quoted to preserve whitespace:

```Makefile
DEST := "$(DESTDIR)/$(PREFIX)"
```

But this doesn't help with the `patsubst` expression, because `patsubst` itself does whitespace folding (and drops leading/trailing whitespace).

Here is one more attempt that seems to work correctly in all cases I can think off:

```Makefile
DEST := "$(shell echo "$(DESTDIR)/$(PREFIX)" | sed 's:///*:/:g')"
```

Samples:

```console
$ make abspath
/usr/local
$ make abspath DESTDIR="/foo bar  baz//xyzzy/"
/foo bar  baz/xyzzy/usr/local
```

Can you see anything wrong with this one?

Reply:
Hi Jan,

that final sentence - not being a native speaker - probably came out wrong!

Please excuse me in case it sounded rude. It wasn't meant to.

> I'm only providing code review feedback. If you disagree, please push back!

At all! On the contrary, I want to thank you for your reviews, which have been very on the point and helpful.

Yes, it's mainly cosmetic 😃 

Such details bother me. I believe you can totally feel me.
But I also agree with you we should put aesthetics in second place in this case. Meaning, try to achieve it only in case correctness it's achieved.

In this context, I believe that `sed s://*:/:g` would make it.

Another option would be `readlink -m "/foo bar  baz//xyzzy/"` but it canonicalizes symlinks (thus I guess it's not what we want, right?).

Reply:
Also, are we fine introducing `sed` as "dependency"?

Reply:
> Please excuse me in case it sounded rude.

Oh no, it didn't. I just wanted to point out that I also am just voicing out my opinions, and I can be wrong or overlook stuff. So if you feel that I'm wrong, please push back and don't just change something because "I want it". Code review is a collaborative process, with the end result hopefully better than what either of the participants would achieve by themselves.

> Another option would be readlink -m "/foo bar baz//xyzzy/"

I would prefer not to do this. To begin with because it doesn't work on macOS. 😄 

> Also, are we fine introducing sed as "dependency"?

I would think so; it is a pretty basic requirement. And I intentionally used an expression that works with the basic pattern support. Normally I would prefer to use `perl` because I know it works the same everywhere (unlike `sed`), but I don't wan't to rely on it being installed.

Reply:
I probably found one last edge (cosmetic) case: DEST to contain a trailing slash. In this case the `uninstall` command looks this way:

```console
PREFIX=/usr/local/ make uninstall
rm -rf \
	"/usr/local//bin/lima" \
	"/usr/local//bin/limactl" \
	"/usr/local//bin/nerdctl.lima" \
	"/usr/local//share/lima" "/usr/local//share/doc/lima"
```

Thus, I'm thinking to define DEST like this:

```makefile
DEST := $(patsubst %/,%,$(shell echo "$(DESTDIR)/$(PREFIX)" | sed 's://*:/:g'))
```

Or with only sed:

```makefile
DEST := $(shell echo "$(DESTDIR)/$(PREFIX)" | sed 's#//*#/#g;s#/*$$##')
```

Results:

```console
PREFIX="/uno/due/foo bar baz//xyzzy/" make prova
/uno/due/foo bar baz/xyzzy
```

```console
PREFIX=/usr/local make prova
/usr/local
```

What do you think? Which one do you prefer? :)

Reply:
> Thus, I'm thinking to define DEST like this:
> 
> ```makefile
> DEST := $(patsubst %/,%,$(shell echo "$(DESTDIR)/$(PREFIX)" | sed 's://*:/:g'))
> ```

As I explained before, `patsubst` will fold multiple spaces into a single space, so should not be used. Maybe it was too subtle, but I intentionally had 2 spaces after `bar`. You dropped the multiple spaces in your test case, so you didn't notice.

 ```
/foo bar  baz
     ^   ^^
````

> Or with only sed:
> 
> ```makefile
> DEST := $(shell echo "$(DESTDIR)/$(PREFIX)" | sed 's#//*#/#g;s#/*$$##')
> ```

Yes, just `sed` is fine. For some reason your command doesn't work for me:

```console
Makefile:3: *** unterminated call to function `shell': missing `)'.  Stop.
```

It has something to do with `#` also being a comment character, and at least "GNU Make 3.81" does not parse it correctly.

You are also still missing the quotes around the right hand side to preserve whitespace.

I've extended my sample, and it worked for me, so I would suggest to use that version (assuming it works for you as well):

```makefile
DEST := "$(shell echo "$(DESTDIR)/$(PREFIX)" | sed 's:///*:/:g; s://*$$::')"
```

(You also dropped one slash from `///*`, which is just a basic pattern way to write `//+`, which only works with `-E`. Of course you can replace a single slash by another identical single slash, but that it kind of pointless. Not that it really matters either way, but since this PR is about aesthetics...)

Reply:
Thank you, Jan!

Yes, I copied and pasted the commands from the shell not from the Makefile I was actually using, silly me 🤦 😄